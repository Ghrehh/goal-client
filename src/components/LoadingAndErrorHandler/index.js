import React, { Component } from 'react';
import PropTypes from 'prop-types';

class LoadingAndErrorHandler extends Component {
  renderError = () => (
    <div>
      <h2>Error</h2>
      <p>{this.props.error.message}</p>
    </div>
  )

  render() {
    if (this.props.loading) return <p>Loading...</p>;
    if (this.props.error) return this.renderError();

    return this.props.children
  }
}

LoadingAndErrorHandler.propTypes = {
  loading: PropTypes.bool.isRequired,
  error: PropTypes.shape({ message: PropTypes.string.isRequired }),
  children: PropTypes.node
};

export default LoadingAndErrorHandler;
