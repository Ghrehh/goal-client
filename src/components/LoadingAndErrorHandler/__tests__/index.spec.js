import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow'
import LoadingAndErrorHandler from '../';

describe('LoadingAndErrorHandler', () => {
  let testContext = {};

  beforeEach(() => {
    testContext = {};
    testContext.props = {
      children: 'foo',
      loading: false,
      error: undefined
    }
  });

  describe('not loading and no error', () => {
    beforeEach(() => {
      const renderer = new ShallowRenderer();
      testContext.result = renderer.render(<LoadingAndErrorHandler {...testContext.props} />);
    });

    it('renders the child component', () => {
      expect(testContext.result).toMatchSnapshot();
    });
  });

  describe('loading', () => {
    beforeEach(() => {
      const renderer = new ShallowRenderer();
      testContext.props.loading = true;
      testContext.result = renderer.render(<LoadingAndErrorHandler {...testContext.props} />);
    });

    it('renders the loading message', () => {
      expect(testContext.result).toMatchSnapshot();
    });
  });

  describe('error', () => {
    beforeEach(() => {
      const renderer = new ShallowRenderer();
      testContext.props.error = { message: 'foobar' }
      testContext.result = renderer.render(<LoadingAndErrorHandler {...testContext.props} />);
    });

    it('renders the loading message', () => {
      expect(testContext.result).toMatchSnapshot();
    });
  });
});
