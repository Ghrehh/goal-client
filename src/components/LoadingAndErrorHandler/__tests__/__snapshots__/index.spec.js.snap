// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoadingAndErrorHandler error renders the loading message 1`] = `
<div>
  <h2>
    Error
  </h2>
  <p>
    foobar
  </p>
</div>
`;

exports[`LoadingAndErrorHandler loading renders the loading message 1`] = `
<p>
  Loading...
</p>
`;

exports[`LoadingAndErrorHandler not loading and no error renders the child component 1`] = `"foo"`;
